File: apps/typescript/app-14-telemedicine/src/services/AuthService.ts
import bcrypt from "bcryptjs";
import { AuthenticatedUser } from "../models/User";
import { AuditEventProducer } from "../mq/AuditEventProducer";
import { UserRepository } from "../repositories/UserRepository";
import { TokenService } from "./TokenService";

export class AuthService {
  constructor(
    private readonly users: UserRepository,
    private readonly tokens: TokenService,
    private readonly auditEvents: AuditEventProducer
  ) {}

  register(username: string, password: string) {
    return this.users.savePatient(username, password);
  }

  login(username: string, password: string) {
    const user = this.users.findByUsername(username);
    if (!user || !bcrypt.compareSync(password, user.passwordHash)) {
      return undefined;
    }

    const payload: AuthenticatedUser = { userId: user.id, username: user.username, role: user.role };
    const token = this.tokens.sign(payload);
    this.auditEvents.publish("auth.login", { userId: user.id });
    return {
      token,
      user: payload
    };
  }

  requireUser(token: string | undefined) {
    return this.tokens.verify(token);
  }
}

File: apps/python/app-01-ecommerce-catalog/services/catalog-service/src/services/auth_service.py
from src.repositories.user_repository import UserRepository


class AuthService:
    def __init__(self, users=None):
        self.users = users or UserRepository()

    def authenticate(self, username, password):
        user = self.users.find_by_credentials(username.strip(), password.strip())
        if not user:
            return None
        return {"id": user["id"], "username": user["username"], "role": user["role"]}

    def profile_for_session(self, session_data):
        user_id = session_data.get("user_id")
        if not user_id:
            return None
        user = self.users.find_by_id(user_id)
        if not user:
            return None
        return {"id": user["id"], "username": user["username"], "role": user["role"]}
